x/dex/keeper/liquidity.go

```diff
@@ -163,23 +165,31 @@ func (k Keeper) Swap(ctx sdk.Context,
 		}
 
 		inAmount, outAmount := liq.Swap(remainingIn, remainingOut)
+		k.SaveLiquidity(ctx, liq)
 
 		remainingIn = remainingIn.Sub(inAmount)
 		totalOut = totalOut.Add(outAmount)
+
+		// break if remainingIn will yield less than 1 tokenOut at current price
+		// this avoids unnecessary iteration since outAmount will always be 0 going forward
+		// this also catches the normal exit case where remainingIn == 0
+		if liq.Price().MulInt(remainingIn).LT(sdk.OneDec()) {
+			orderFilled = true
+			break
+		}
+
 		if useMaxOut {
```

Comment:
Do we need to gate this? Seems like we only need to gate whether remainingOut is used to break early.